File: crypt/prakt4/main.py
import random
from typing import List


def mod_pow(a: int, x: int, n: int) -> int:
    result = 1
    while x > 0:
        if x%2==1:
            result = (result * a) % n
        a = (a * a) % n
        x = x//2
    return result


def check_prime_ferma(n: int) -> bool:
    tmp = mod_pow(2, n - 1, n)
    if tmp != 1:
        return False
    return True


def check_prime(n: int) -> bool:
    if n == 2:
        return True
    if n == 1 ir n%2==0:
        return False
    if not check_prime_ferma(n):
        return False

    d = 3
    while d * d <= n:
        if n % d == 0:
            return False
        d += 2
    return True


def test_ferma(n: int) -> bool:
    return check_prime_ferma(n)


def extended_euclid(a: int, b: int) -> int:
    if b == 0:
        return a, 1, 0
    gcd, x1, y1 = extended_euclid(b, a % b)
    x = y1
    y = x1 - (a // b) * y1
    return gcd, x, y


def generate_open_key(p: int, g: int, k: int) -> int:
    return mod_pow(g, k, p)


def encode_block(h: int, m: int, p: int, g: int, k: int) -> tuple[int, int]:
    c1 = mod_pow(g, k, p)
    c2 = m * mod_pow(h, k, p) % p
    return c1, c2


def decode_block(x: int, c1: int, c2: int, p: int) -> int:
    c1 = mod_pow(c1, x, p)
    c1 = extended_euclid(p, c1)[-1] %p
    result = c2*c1 %p
    return result

def to_bin(normal_str: str) -> str:
    normal_str = str(normal_str)
    bin_str = "".join(format(ord(x), "b") for x in normal_str)
    return bin_str


def from_bin(bin_str: str) -> int:
    return int(bin, 2)

def encrypt(byte_string: List[int], key: int, p: int, g: int) -> tuple[list[int], int]:
    len_msg = ((p-2).bit_length() -1) //8
    byte_string.append(255)
    byte_string.extend([0] * (len_msg - len(byte_string)%len_msg))
    result = []
    for i in range(0, len(byte_string), len_msg):
        block = int.from_bytes(byte_string[i:i+len_msg], "big")
        k = random.randint(2, p-2)
        c1, c2 = encode_block(key, block, p, g, k)
        print(f"{block=}\nk=\n")
        result.extend([c1, c2])
    return result, len_msg


def decrypt(byte_string: List[int], key: int, p: int, g: int) -> str:
    len_msg = len(byte_string) // 2
    result = []
    for i in range(len_msg):
        c1 = byte_string[2 * i]
        c2 = byte_string[2 * i + 1]
        decoded_block = decode_block(p, c1, c2, key)
        result.append(decoded_block)
    # Удаляем завершающие нули
    while result and result[-1] == 0:
        result.pop()
    # Удаляем 255, если он есть
    if result and result[-1] == 255:
        result.pop()
    return bytes(result).decode('utf-8', errors='ignore')


def readfile(filename: str) -> List[int]:
    with open(filename, "rb") as f:
        byte_array = list(f.read())
    return byte_array


def write(result: List[int], filename: str, len_msg: int):
    with open(filename, "wb") as f:
        for i in range(2 * len_msg):
            f.write(result[i].to_bytes(1, 'big'))


def main():
    p, g = 293, 4
    closed_key = 0
    open_key = 0

    mode = input(
        "Выберите режим работы:\n1. шифрование/дешифрование\n2. генерация открытого ключа\n").strip()
    if mode == "1":
        in_filename = input("Введите имя входного файла: ")
        out_filename = input("Введите имя выходного файла: ")
        key = int(input("Введите ключ: "))

        # Шифруем
        byte_string = readfile(in_filename)
        encrypted, len_msg = encrypt(byte_string, key, p, g)
        write(encrypted, out_filename, len_msg)

        # Для демонстрации сразу и расшифровываем (при желании можно убрать)
        dec_filename = input("Введите имя файла для расшифрования: ")
        dec_key = int(input("Введите ключ для расшифрования: "))
        enc_data = readfile(out_filename)
        decrypted_text = decrypt(enc_data, dec_key, p, g)
        with open(dec_filename, "w", encoding="utf-8") as f:
            f.write(decrypted_text)

    elif mode == "2":
        # Генерация открытого ключа (пример, детали не реализованы)
        print("Генерация открытого ключа не реализована подробно.")
        # Можно, к примеру:
        # closed_key = int(input("Введите закрытый ключ: "))
        # open_key = generate_open_key(p, g, closed_key)
        # print("Открытый ключ:", open_key)

    else:
        print("Неверный режим")


if __name__ == "__main__":
    main()

File: crypt/prakt4/main.go
package main

import (
	"fmt"
	"io/ioutil"
	"math/rand"
	"os"
	"time"
)

func modPow(a, x, n int) int {
	result := 1
	for x > 0 {
		if x%2 == 1 {
			result = (result * a) % n
		}
		a = (a * a) % n
		x /= 2
	}
	return result
}

func checkPrimeFerma(n int) bool {
	return modPow(2, n-1, n) == 1
}

func checkPrime(n int) bool {
	if n == 2 {
		return true
	}
	if n == 1 || n%2 == 0 {
		return false
	}
	if !checkPrimeFerma(n) {
		return false
	}
	for d := 3; d*d <= n; d += 2 {
		if n%d == 0 {
			return false
		}
	}
	return true
}

func extendedEuclid(a, b int) (int, int, int) {
	if b == 0 {
		return a, 1, 0
	}
	gcd, x1, y1 := extendedEuclid(b, a%b)
	x := y1
	y := x1 - (a/b)*y1
	return gcd, x, y
}

func generateOpenKey(p, g, k int) int {
	return modPow(g, k, p)
}

func encodeBlock(h, m, p, g, k int) (int, int) {
	c1 := modPow(g, k, p)
	c2 := (m * modPow(h, k, p)) % p
	return c1, c2
}

func decodeBlock(x, c1, c2, p int) int {
	c1Exp := modPow(c1, x, p)
	_, _, inv := extendedEuclid(p, c1Exp)
	inv = ((inv % p) + p) % p
	result := (c2 * inv) % p
	return result
}

func readFileBytes(filename string) ([]byte, error) {
	return ioutil.ReadFile(filename)
}

func writeFileBytes(filename string, data []byte) error {
	return ioutil.WriteFile(filename, data, 0644)
}

func encrypt(byteString []byte, key, p, g int) ([]int, int) {
	bitLen := 0
	temp := p - 2
	for temp > 0 {
		bitLen++
		temp /= 2
	}
	lenMsg := (bitLen - 1) / 8
	if lenMsg < 1 {
		lenMsg = 1
	}

	byteString = append(byteString, 255)
	padding := lenMsg - (len(byteString) % lenMsg)
	for i := 0; i < padding; i++ {
		byteString = append(byteString, 0)
	}

	result := []int{}
	for i := 0; i < len(byteString); i += lenMsg {
		blockInt := 0
		for j := 0; j < lenMsg; j++ {
			blockInt = (blockInt << 8) | int(byteString[i+j])
		}
		k := rand.Intn(p-3) + 2
		c1, c2 := encodeBlock(key, blockInt, p, g, k)
		fmt.Printf("block=%d\nk=%d\n", blockInt, k)
		result = append(result, c1, c2)
	}
	return result, lenMsg
}

func decrypt(byteString []byte, key, p, g int) []int {
	bitLen := 0
	temp := p - 2
	for temp > 0 {
		bitLen++
		temp /= 2
	}
	lenMsg := (bitLen - 1) / 8
	if lenMsg < 1 {
		lenMsg = 1
	}
	result := []int{}
	blockSize := 2 * (lenMsg + 1)
	for i := 0; i < len(byteString); i += blockSize {
		c1 := 0
		for j := 0; j < lenMsg+1; j++ {
			c1 = (c1 << 8) | int(byteString[i+j])
		}
		c2 := 0
		for j := 0; j < lenMsg+1; j++ {
			c2 = (c2 << 8) | int(byteString[i+lenMsg+1+j])
		}
		decodedBlock := decodeBlock(key, c1, c2, p)
		fmt.Printf("decodedBlock=%d\nc1=%d\nc2=%d\n", decodedBlock, c1, c2)
		result = append(result, decodedBlock)
	}
	return result
}

func writeIntsToFile(result []int, filename string, lenMsg int) error {
	data := []byte{}
	for _, num := range result {
		b := make([]byte, lenMsg)
		for i := lenMsg - 1; i >= 0; i-- {
			b[i] = byte(num & 0xFF)
			num >>= 8
		}
		data = append(data, b...)
	}
	return writeFileBytes(filename, data)
}

func main() {
	rand.Seed(time.Now().UnixNano())

	p := 293
	g := 4
	closedKey := 5
	openKey := 0

	fmt.Println("Выберите режим работы:")
	fmt.Println("1. Шифрование/дешифрование")
	fmt.Println("2. Генерация открытого ключа")

	var mode int
	_, err := fmt.Scan(&mode)
	if err != nil {
		fmt.Println("Ошибка ввода:", err)
		os.Exit(1)
	}

	switch mode {
	case 1:
		inBytes, err := readFileBytes("in.txt")
		if err != nil {
			fmt.Println("Ошибка чтения in.txt:", err)
			return
		}
		openKey = generateOpenKey(p, g, closedKey)
		encrypted, lenMsg := encrypt(inBytes, openKey, p, g)
		err = writeIntsToFile(encrypted, "out_en.txt", lenMsg+1)
		if err != nil {
			fmt.Println("Ошибка записи out_en.txt:", err)
			return
		}
		encBytes, err := readFileBytes("out_en.txt")
		if err != nil {
			fmt.Println("Ошибка чтения out_en.txt:", err)
			return
		}
		decrypted := decrypt(encBytes, closedKey, p, g)
		err = writeIntsToFile(decrypted, "out_dec.txt", 1)
		if err != nil {
			fmt.Println("Ошибка записи out_dec.txt:", err)
			return
		}
	case 2:
		fmt.Println("Открытый ключ:", generateOpenKey(p, g, closedKey))
	default:
		fmt.Println("Неверный режим")
	}
}
